### ISSUE TITLE
deadlock on shutdown

### ISSUE THREAD

**Issue by skriss** [projectcontour/contour#4163]
https://github.com/projectcontour/contour/runs/4113809841?check_suite_focus=true#step:6:790
https://github.com/projectcontour/contour/runs/4120027004?check_suite_focus=true#step:6:685

Hypothesis is that the sigterm handler hasn't actually started by the time Ginkgo sends a sigterm to shut down Contour, so it's never handled and the test times out waiting for Contour to shut down.

**Comment by tsaarni** [comment]
I also bumped into this previously. It seemed to handle the signal

```
    time="2021-11-05T05:31:00Z" level=info msg="shutting down" context=sigterm-handler signal=terminated
    time="2021-11-05T05:31:00Z" level=error msg="terminated HTTP server with error" context=metricsvc error="http: Server closed"
    time="2021-11-05T05:31:00Z" level=info msg="stopped event handler" context=contourEventHandler
    time="2021-11-05T05:31:00Z" level=error msg="terminated HTTP server with error" context=debugsvc error="http: Server closed"
```

but still it took the full 60 seconds after that...

**Comment by tsaarni** [comment]
There is difference in successful and unsuccessful shutdown:

When shutdown is successful, xDS server goroutine has reached this far (see successful log [here](https://github.com/projectcontour/contour/runs/4120027004?check_suite_focus=true#step:6:142)) https://github.com/projectcontour/contour/blob/cb3ec9aee1ea74ac30b8823ab0b3b03588dfb3d8/cmd/contour/serve.go#L571

In case of unsuccessful shutdown it seems like that goroutine is stuck here https://github.com/projectcontour/contour/blob/cb3ec9aee1ea74ac30b8823ab0b3b03588dfb3d8/cmd/contour/serve.go#L567-L568

Why would that get stuck?

If that is the case, then the `WaitGroup` is stuck forever and shutdown does not proceed, since one of the goroutines never calls `Done()` https://github.com/projectcontour/contour/blob/cb3ec9aee1ea74ac30b8823ab0b3b03588dfb3d8/internal/workgroup/group.go#L92-L99

**Comment by skriss** [comment]
:+1: agreed, that looks suspicious. Haven't been able to spot the bug there yet.

One other suspect thing I'm wondering about: https://github.com/projectcontour/contour/blob/main/cmd/contour/serve.go#L250-L253

The controller-runtime manager is setting up its own SIGTERM handler, rather than using the workgroup context / relying on the Contour SIGTERM handler. Not sure yet if that could be causing any issues.

**Comment by tsaarni** [comment]
It seems to be OK according to https://pkg.go.dev/os/signal#Notify

> It is allowed to call Notify multiple times with the same channel: each call expands the set of signals sent to that channel. The only way to remove signals from the set is to call Stop.
> 
> It is allowed to call Notify multiple times with different channels and the same signals: each channel receives copies of incoming signals independently.
>

**Comment by skriss** [comment]
The test flake was resolved in #4164 but there's still a deadlock in here on shutdown.

**Comment by skriss** [comment]
Seems possible that this has been resolved through the work @sunjayBhatia did moving everything to controller-runtime runnables, closing out but can reopen if needed.
